lib/IRGen/GenFunc.cpp

```diff
@@ -719,6 +720,88 @@ static unsigned findSinglePartiallyAppliedParameterIndexIgnoringEmptyTypes(
+  switch (fieldConvention) {
+  case ParameterConvention::Indirect_In:
+  case ParameterConvention::Indirect_In_Constant: {
+
+    auto initStackCopy = [&addressesToDeallocate, &needsAllocas, &param,
+                          &subIGF](const TypeInfo &fieldTI, SILType fieldTy,
+                                   Address fieldAddr) {
+      // The +1 argument is passed indirectly, so we need to copy into a
+      // temporary.
+      needsAllocas = true;
+      auto stackAddr = fieldTI.allocateStack(subIGF, fieldTy, "arg.temp");
+      auto addressPointer = stackAddr.getAddress().getAddress();
+      fieldTI.initializeWithCopy(subIGF, stackAddr.getAddress(), fieldAddr,
+                                 fieldTy, false);
+      param.add(addressPointer);
+
+      // Remember to deallocate later.
+      addressesToDeallocate.push_back(
+          AddressToDeallocate{fieldTy, fieldTI, stackAddr});
+    };
+
+    if (needsLoad) {
+      // If the closure is [onstack] it only captured the address of the
+      // value. Load that address from the context.
+      Explosion addressExplosion;
+      cast<LoadableTypeInfo>(fieldTI).loadAsCopy(subIGF, fieldAddr,
+                                                 addressExplosion);
+      assert(fieldTy.isAddress());
+      auto newFieldTy = fieldTy.getObjectType();
+      auto &newFieldTI =
+          subIGF.getTypeInfoForLowered(newFieldTy.getASTType());
+      fieldAddr =
+          newFieldTI.getAddressForPointer(addressExplosion.claimNext());
+      initStackCopy(newFieldTI, newFieldTy, fieldAddr);
+    } else {
+      initStackCopy(fieldTI, fieldTy, fieldAddr);
+    }
+    break;
+  }
+  case ParameterConvention::Indirect_In_Guaranteed:
+    if (needsLoad) {
+      cast<LoadableTypeInfo>(fieldTI).loadAsCopy(subIGF, fieldAddr, param);
+    } else {
+      // The argument is +0, so we can use the address of the param in
+      // the context directly.
+      param.add(fieldAddr.getAddress());
+      dependsOnContextLifetime = true;
+    }
+    break;
+  case ParameterConvention::Indirect_Inout:
+  case ParameterConvention::Indirect_InoutAliasable:
+    // Load the address of the inout parameter.
+    cast<LoadableTypeInfo>(fieldTI).loadAsCopy(subIGF, fieldAddr, param);
+    break;
+  case ParameterConvention::Direct_Guaranteed:
+  case ParameterConvention::Direct_Unowned:
+    // If the type is nontrivial, keep the context alive since the field
+    // depends on the context to not be deallocated.
+    if (!fieldTI.isPOD(ResilienceExpansion::Maximal))
+      dependsOnContextLifetime = true;
+
+    // Load these parameters directly. We can "take" since the parameter is
+    // +0. This can happen since the context will keep the parameter alive.
+    cast<LoadableTypeInfo>(fieldTI).loadAsTake(subIGF, fieldAddr, param);
+    break;
+  case ParameterConvention::Direct_Owned:
+    // Copy the value out at +1.
+    cast<LoadableTypeInfo>(fieldTI).loadAsCopy(subIGF, fieldAddr, param);
```

Comment:
I think this should be a take for async context since the caller of the async function guarantees the +1.
Also I assume we never actually hit any of the indirect cases (except for indirect_inout) for async contexts. Because they all should map to a CanInout(AST.Type). Can we assert this? and prove me right or wrong.

Reply:
Basically the rule I have in mind is: The SwiftContext does not take ownership of any arguments or result. The ownership is held by the allocated (alloc_stack) memory location in the caller or by the reference passed by the caller. Therefore all indirect arguments should be just passing the address allocated in the caller and direct arguments should just extract the value out of the SwiftContext at +0.

Reply:
Thanks again for the thorough explanation!  These functions are now removed and moving values into and out of the context generate no ARC traffic.